aptos-move/move-examples/rewards_pool/sources/rewards_pool.move

```diff
@@ -0,0 +1,211 @@
+                vector::push_back(&mut non_empty_reward_tokens, reward_token);
+                vector::push_back(&mut reward_per_tokens, reward);
+            };
+        });
+        (non_empty_reward_tokens, reward_per_tokens)
+    }
+
+    /// Allow a claimer to claim the rewards for a past epoch.
+    public fun claim_rewards(
+        claimer: address,
+        rewards_pool: Object<RewardsPool>,
+        reward_token: Object<Metadata>,
+        epoch: u64,
+    ): FungibleAsset acquires RewardsPool {
+        assert!(epoch < epoch::now(), EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH);
+        let reward = rewards(claimer, rewards_pool, reward_token, epoch);
+        assert!(reward > 0, ENO_REWARDS_TO_CLAIM);
+        let reward_store = smart_table::borrow(&safe_rewards_pool_data(&rewards_pool).epoch_rewards, reward_token);
+        let store_signer = &object::generate_signer_for_extending(&reward_store.store_extend_ref);
+        fungible_asset::withdraw(store_signer, reward_store.store, reward)
+    }
+
+    /// Add rewards to the specified rewards pool.
+    public fun add_rewards(
+        rewards_pool: Object<RewardsPool>,
+        fungible_assets: vector<FungibleAsset>,
+        epoch: u64,
+    ) acquires RewardsPool {
+        vector::for_each(fungible_assets, |fa| {
+            let reward_token = fungible_asset::metadata_from_asset(&fa);
+            // This aborts if the reward token is not supported.
+            let reward_store = smart_table::borrow_mut(
+                &mut unchecked_mut_rewards_pool_data(&rewards_pool).epoch_rewards,
+                reward_token,
+            );
+            let amount = fungible_asset::amount(&fa);
+            let current_amount = smart_table::borrow_mut_with_default(&mut reward_store.amount, epoch, 0);
+            *current_amount = *current_amount + amount;
+            fungible_asset::deposit(reward_store.store, fa);
+        });
+    }
+
+    /// This should only be called by system modules to increase the shares of a claimer.
+    public(friend) fun increase_allocation(
+        claimer: address,
+        rewards_pool: Object<RewardsPool>,
+        amount: u64,
+    ) acquires RewardsPool {
+        pool_u64::buy_in(unchecked_mut_epoch_allocation(&rewards_pool), claimer, amount);
+    }
+
+    /// This should only be called by system modules to decrease the shares of a claimer.
+    public(friend) fun decrease_allocation(
+        claimer: address,
+        rewards_pool: Object<RewardsPool>,
+        amount: u64,
+    ) acquires RewardsPool {
+        let epoch_allocations = &mut unchecked_mut_rewards_pool_data(&rewards_pool).epoch_allocations;
+        let current_epoch = epoch::now();
+        let pool = smart_table::borrow_mut(epoch_allocations, current_epoch);
+        pool_u64::redeem_shares(pool, claimer, (amount as u128));
+
+        // Delete the epoch allocation pool if there are no remaining claimers.
+        if (pool_u64::total_coins(pool) == 0) {
```

Comment:
Does `pool_u64::total_coins(pool) == 0` always equal to `pool_u64::total_shares(pool) == 0`? I would suggest to add a function in pool module called `is_empty` and assert there.

Reply:
Yes its the same because total coins is never changed. Its always equal to total shares

Reply:
total coins = total shares? sry I don't get this part.